bigtable/metrics.go

```diff
@@ -0,0 +1,401 @@
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create attempt_latencies
+	tf.attemptLatencies, err = meter.Float64Histogram(
+		metricNameAttemptLatencies,
+		metric.WithDescription("Client observed latency per RPC attempt."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create server_latencies
+	tf.serverLatencies, err = meter.Float64Histogram(
+		metricNameServerLatencies,
+		metric.WithDescription("The latency measured from the moment that the RPC entered the Google data center until the RPC was completed."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create retry_count
+	tf.retryCount, err = meter.Int64Counter(
+		metricNameRetryCount,
+		metric.WithDescription("The number of additional RPCs sent after the initial attempt."),
+	)
+	return err
+}
+
+// builtinMetricsTracer is created one per operation
+// It is used to store metric instruments, attribute values
+// and other data required to obtain and record them
+type builtinMetricsTracer struct {
+	ctx            context.Context
+	builtInEnabled bool
+
+	// attributes that are specific to a client instance and
+	// do not change across different operations on client
+	clientAttributes []attribute.KeyValue
+
+	instrumentOperationLatencies metric.Float64Histogram
+	instrumentServerLatencies    metric.Float64Histogram
+	instrumentAttemptLatencies   metric.Float64Histogram
+	instrumentRetryCount         metric.Int64Counter
+
+	tableName   string
+	method      string
+	isStreaming bool
+
+	currOp opTracer
+}
+
+type opTracer struct {
+	attemptCount int64
+
+	startTime time.Time
+
+	// gRPC status code of last completed attempt
+	status string
+
+	currAttempt attemptTracer
+}
+
+func (o *opTracer) setStartTime(t time.Time) {
+	o.startTime = t
+
+}
+
+func (o *opTracer) setStatus(status string) {
+	o.status = status
+}
+
+func (o *opTracer) incrementAttemptCount() {
+	o.attemptCount++
+}
+
+type attemptTracer struct {
+	startTime time.Time
+	clusterID string
+	zoneID    string
+
+	// gRPC status code
+	status string
+
+	// Server latency in ms
+	serverLatency float64
+
+	// Error seen while getting server latency from headers
+	serverLatencyErr error
+}
+
+func (a *attemptTracer) setStartTime(t time.Time) {
+	a.startTime = t
+}
+
+func (a *attemptTracer) setClusterID(clusterID string) {
+	a.clusterID = clusterID
+}
+
+func (a *attemptTracer) setZoneID(zoneID string) {
+	a.zoneID = zoneID
+}
+
+func (a *attemptTracer) setStatus(status string) {
+	a.status = status
+}
+
+func (a *attemptTracer) setServerLatency(latency float64) {
+	a.serverLatency = latency
+}
+
+func (a *attemptTracer) setServerLatencyErr(err error) {
+	a.serverLatencyErr = err
+}
+
+func (tf *builtinMetricsTracerFactory) newBuiltinMetricsTracer(ctx context.Context, tableName string, isStreaming bool) builtinMetricsTracer {
```

Comment:
Since this is a factory method, could you rename the `new` part `get` or `create`?

Reply:
Changed to create